versiebeheer.md

```diff
@@ -0,0 +1,155 @@
+# Versie beheer en release management
+
+## API versies
+
+Versiebeheer is gebaseerd op [Semantic Versioning](https://semver.org) en de
+[API strategie](https://aandeslagmetdeomgevingswet.nl/digitaal-stelsel/technisch-aansluiten/standaarden/api-uri-strategie/).
+Elke  API heeft een eigen versienummer, opgebouwd uit `MAJOR.MINOR.PATCH` 
+element. Bijvoorbeeld versie `2.1.8`:
+
+Gegeven dit versienummer, elke ophoging van de:
+
+* `MAJOR` versie betreft een *backwards-incompatible* API wijziging,
+* `MINOR` versie betreft het toevoegen van een *backwards-compatible* API 
+  wijziging,
+* `PATCH` versie betreft het oplossen van *backwards-compatible* issues.
+
+Verder wordt een aantal labels gebruikt, achter het versienummer:
+
+* Als er geen aanduiding bij het versienummer staat, gaat het altijd om de
+  meest actuele productie versie van de API.
+* `supported` versies zijn oudere maar ondersteunde versies van de API,
+* `archived` versies zijn oudere, niet meer ondersteunde versies van de API.
+
+Voor meer interne doeleinden kennen we tenslotte nog de volgende labels achter
+het versienummer:
+
+* `RC` (Release Candidate) als de API versie feature-complete is. Alleen
+  kritieke issues worden in deze versie nog verwerkt,
+* `beta` als de versie gereed is voor testers en er geen grote wijzigingen meer
+  verwacht worden. Alleen essentiele zaken worden in dit stadium nog 
+  toegevoegd,
+* `alpha` als de versie nog in ontwikkeling is en er geen enkele garantie is
+  over de werking, stabiliteit e.d.
+
+### Voorbeelden van backwards-incompatible wijzigingen
+
+* Wijzigingen in de URL of fundamentale wijzigingen in de request/response
+  schema's die bij een resource horen,
+* Er is een data migratie nodig om van de vorige naar de nieuwe versie te komen.
+  Bijvoorbeeld omdat een attribuut wordt opgesplitst in meerdere attributen, of
+  juist worden samengevoegd,
+* Verwijderen, hernoemen of type-wijziging van een attribuut, query parameter
+  of header,
+* Verwijderen of hernoemen van een resource of API als geheel,
+* Toevoegen van een verplicht attribuut of header.
+
+**LET OP: We spreken altijd over backwards-incompatible wijzigingen vanuit het 
+provider perspectief! Een provider met API op versie `2.1.8` kan altijd goed
+communiceren met een consumer die `2.0.0` t/m `2.1.8` praat (kortgezegd,
+`2.x`) maar het is niet persé waar dat een consumer op versie `2.2.0` goed kan 
+communiceren met een API op versie `2.1.8`.**
+
+### Voorbeelden van backwards-compatible wijzigingen:
+
+* Toevoegen van attributen die optioneel zijn, `null` mogen zijn of een andere
+  standaard waarde hebben.
+* Toevoegen van een element aan een enumeratie,
```

Comment:
Ik twijfel er aan of dit backwards compatible is. Stel er wordt een response gegeven aan een Consumer met een enumeratie waarde die in de API versie die de Consumer gebruikt niet voorkomt. Volgens mij levert dat dan compatibiliteits problemen op.

Reply:
Ik heb dit punt overgenomen met wat Microsoft als compatible zag:

> Addition of a member to an enumeration

https://docs.microsoft.com/en-us/graph/versioning-and-support

Mij lijkt het geen probleem.

Reply:
Het punt van Robert kan ik me wel voorstellen. Waar op zich wel dingen mis kunnen gaan is wanneer bepaalde overgangen van waarden vastgelegd zijn.
Dus bijvoorbeeld tussen de statussen "Verzoek ontvangen" en "In behandeling genomen" komt ineens de status "Ter beoordeling". Dat hoeft geen probleem te zijn maar wanneer het niet toegestaan wordt om van "Verzoek ontvangen" direct naar "In behandeling genomen" te gaan gaat wel mis.

Ik denk eigenlijk dat alles wat gegevens raakt backwards in-compatible is. Dus wijzigingen aan resources en/of attributen en/of relaties maar ook enumeraties. Dat raakt de inhoud en validatie.

Reply:
Ok maar je betrekt er nu business logica bij die wijzigt. Dat is meer dan een element toevoegen.

Ik wil em wel weghalen hoor. Het zijn maar voorbeelden.

Reply:
In de standaard zit meer dan alleen het technisch uitwisselen van berichten. Bijvoorbeeld autorisatie of het gedrag dat een DRC een relatie tussen een zaak en een informatieobject ook zelf, onder water, vastlegt in een ZRC.

Prima wat mij betreft maar dan moeten we gedrag en business logica ook meenemen in versiebeheer etc.

Of we accepteren alleen puur berichten en berichtuitwisseling in beheer maar dat is denk ik een stap terug die we nou juist niet willen.

Reply:
@Joeri, Ik wil dit document graag toevoegen en de pull request dus ook z.s.m. goedkeuren.
Doe me een lol en voeg een aantekening toe waarmee we de bovenstaande vraag even in het document parkeren. Daar kijken we dan later nog even naar. Dan kan ik dit PR goedkeuren en kan jij het mergen met de master.

Reply:
Kan jij met een `suggestion` aangeven wat je er in wilt?

![image](https://user-images.githubusercontent.com/96970/52870982-6053f780-3149-11e9-831d-232d82e39b40.png)